chore(maintenance): migrate logger utility to biome

daschaa opened the pull request:
## Summary

### Changes

Adds biome to the logger package. Closes #2798 

<!-- What is this PR solving? Write a clear description or reference the issue(s) it addresses. -->

**Issue number: 2798** 

<!-------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/CONTRIBUTING.md#sending-a-pull-request
- Check that there isn't already a PR that addresses the same issue. If you find a duplicate, please leave a comment under the existing PR so we can discuss how to move forward
- Check that the change meets the project's tenets https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets
- Add a PR title that follows the conventional commit semantics - https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2
- If relevant, add tests that prove that the change is effective and works
- Whenever relevant, make sure to comment functions/methods/types and make appropriate changes to the documentation
------->

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

daschaa commented on packages/logger/src/types/Log.ts:
```diff
@@ -1,11 +1,11 @@
 import type { EnvironmentVariablesService } from '../config/EnvironmentVariablesService.js';
+import type { LogLevel as LogLevelList } from '../constants.js';
```
The linter complained about LogLevel being declared twice, so I decided to give the import an alias in this file. Please let me know if you are happy with this or if it should be changed.

dreamorosi replied:
I think it's fine, good call!

dreamorosi commented on packages/logger/tests/unit/Logger.test.ts:
```diff
@@ -1,25 +1,25 @@
+import type { LambdaInterface } from '@aws-lambda-powertools/commons/types';
 /**
```
Could you please put the comment currently in L2 to L6 at the very top of the file? It's needed for Jest grouping, and it got moved by the Biome formatting.

daschaa replied:
Done :)

dreamorosi reviewed (changes requested):
Thank you for working on this!

I've just left one comment and then it's good to merge :)

dreamorosi commented:
@daschaa - Could you also take a look at the SonarCloud finding [here](https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=2813&id=aws-powertools_powertools-lambda-typescript&open=AZDfnaK1NKxyLq6_nzFi)?

I'll check your other two comments in the meanwhile - need to clone the repo to answer definitively.

dreamorosi commented:
When you can, could you also pls leave a comment under the linked issue (& all others you're working or want to pick up) so that I can assign them to you? Thanks!

dreamorosi commented on packages/logger/src/Logger.ts:
```diff
@@ -697,9 +692,11 @@ class Logger extends Utility implements LoggerInterface {
     const references = new WeakSet();
 
     return (key, value) => {
+      // biome-ignore lint/style/noParameterAssign:
```
I've pulled the branch and tested this, if we create a local var & use it throughout like this:

```ts
return (key, value) => {
  let replacedValue = value;
  if (this.#jsonReplacerFn) replacedValue = this.#jsonReplacerFn?.(key, replacedValue);

  if (replacedValue instanceof Error) {
    replacedValue = this.getLogFormatter().formatError(replacedValue);
  }
  if (typeof replacedValue === 'bigint') {
    return replacedValue.toString();
  }
  if (typeof replacedValue === 'object' && replacedValue !== null) {
    if (references.has(replacedValue)) {
      return;
    }
    references.add(replacedValue);
  }

  return replacedValue;
};
```

it's fine and the tests stay green - I'd say let's do something like this so we can remove the Biome suppression.

daschaa commented:
@dreamorosi I'll check the sonarcloud issue and comment on the issues i'd like to pick :) Thank you for your support ❤️ Really awesome opportunity for me to learn about biome and to (hopefully) help you a bit out here :)

daschaa commented:
@dreamorosi Thanks for the review! 
- I have put the grouping comment to the top of each test file (btw I did not know about this feature in jest, really nice)
- I have removed the suppression of the variable override and introduced the local variable like you suggested
- I have fixed the SonarCloud finding by changing from a traditional for loop to a for...of loop

dreamorosi reviewed (approved):
Thank you for addressing all the feedback and for helping improve the project!
